### PR TITLE
Issue 317 of front end Fixed

### PR DESCRIPTION
**Pull request by Gauravseta**
Identation is proper because no IDE has been used for identation as github editor is used.

Logic flow:
1.From Front end locale value("en"/"es") is inserted in params,which comes at backend in Mentor/Mentee Controller.java
2. From the entity we received at backend, fetch value ("en"/"es") , Using mentor/mentee service  get respective Locale Object by passing value ("en"/"es").
3.Locale Object is to be set in entity now.
4.Mentor/Mentee table is modified accordingly, 'locale_id' column is added with FK to locale table,
Alter scripts will be commented below.
5. InstitutionUser.java has been updated and added Locale field representing locale_id in mentor/mentee table.

### CODE REVIEW

**Comment by Gauravseta** [comment]
How to update this PR?

**Comment by Gauravseta** [comment]
ddl.sql has been modified, locale_id column is added in mentee and mentor table.
Alter table script for changes made is :
/** For Mentor table **/
ALTER TABLE mentor
ADD COLUMN locale_id BIGINT NOT NULL
ADD CONSTRAINT m_li_fk
FOREIGN KEY (locale_id)
REFERENCES locale (id)
ON DELETE CASCADE
ON UPDATE NO ACTION

/** ends**/

/** for Mentee table**/
ALTER TABLE mentee
ADD COLUMN locale_id BIGINT NOT NULL
CONSTRAINT me_li_fk
FOREIGN KEY (locale_id)
REFERENCES locale (id)
ON DELETE CASCADE
ON UPDATE NO ACTION
/** ends **/

File: src/main/java/com/livingprogress/mentorme/controllers/MenteeController.java [hunk 1]
```diff
@@ -177,6 +178,17 @@ public Mentee create(@RequestBody Mentee entity) throws MentorMeException {
                 break;
             }
         }
+	    
+        //set the locale object
+        Locale localeObj = entity.getLocale();
+	 if(localeObj!=null) {
```
Comment (chq-matteo) [inline src/main/java/com/livingprogress/mentorme/controllers/MenteeController.java]: I think there should be a space after `if` as in other parts of the code
(e.g. `Helper.java` line 1254)

Comment (Gauravseta) [inline src/main/java/com/livingprogress/mentorme/controllers/MenteeController.java]: thanks.

**Comment by huangqun** [comment]
@Gauravseta since the locale_id column is required (cannot be null), what value will the existing users have? Doesn't look like your alter script handles that properly?

**Comment by Gauravseta** [comment]
Give me 10 minutes

**Comment by Gauravseta** [comment]
@huangqun  Default will be '1' indicating english language.

**Comment by Gauravseta** [comment]
/** For Mentor table **/
ALTER TABLE mentor
ADD COLUMN locale_id BIGINT NOT NULL default '1'
ADD CONSTRAINT m_li_fk
FOREIGN KEY (locale_id)
REFERENCES locale (id)
ON DELETE CASCADE
ON UPDATE NO ACTION

/** ends**/

/** for Mentee table**/
ALTER TABLE mentee
ADD COLUMN locale_id BIGINT NOT NULL default '1'
CONSTRAINT me_li_fk
FOREIGN KEY (locale_id)
REFERENCES locale (id)
ON DELETE CASCADE
ON UPDATE NO ACTION
/** ends **/

**Comment by huangqun** [comment]
@Gauravseta the alter script doesn't work:
image

**Comment by Gauravseta** [comment]
Its was running just 1 minute.

**Comment by Gauravseta** [comment]
@huangqun updated script is : 
For mentor
/** starts **/
ALTER TABLE mentor
ADD COLUMN locale_id BIGINT NOT NULL default '1',
ADD CONSTRAINT m_li_fk
FOREIGN KEY (locale_id)
REFERENCES locale (id)
ON DELETE CASCADE
ON UPDATE NO ACTION
/** ends **/


For menteee
/**starts**/
ALTER TABLE mentee
ADD COLUMN locale_id BIGINT NOT NULL default '1',
ADD CONSTRAINT me_li_fk
FOREIGN KEY (locale_id)
REFERENCES locale (id)
ON DELETE CASCADE
ON UPDATE NO ACTION
/***ends**/

**Comment by Gauravseta** [comment]
Copy paste the above script under the tags start and end.It will run and add column in both.

**Comment by huangqun** [comment]
Did you run `mvn clean test` with your updated code? I keep seeing a `Column count doesn't match value count at row 1` message when running the tests.

**Comment by huangqun** [comment]
And the final result is:
Tests run: 130, Failures: 0, Errors: 130, Skipped: 0

We cannot accept this, please fix it and re-submit.

**Comment by Gauravseta** [comment]
This issue exist because on writing alter script, column is added at last in table,
where as i have directly added column at 2nd place in mentor and mentee table. thats why when it is saying column mismatch

**Comment by Gauravseta** [comment]
@huangqun  I dont know even I am getting same error now,
 Can you please confirm me the steps which are to be followed,
I followed these steps,
1.schema.sql , mentorme and mentormetest db created.
2.ddl.sql 
3.I ran the above alter script.
4. testdata.sql
5. mvn clean test

It says same thing to me,Same error as u,but it work correctly in actual deployment.

I googled this above error , majority people says its because ,
Insert into statement has issues, column in missed whereas we are adding a column and then running testcases so I guess it will fail, Dont know how to write test cases but surely its simple mismatch which we are adding but testcases is not made to support this.

**Comment by Gauravseta** [comment]
M sure something is missing in test cases,Are testcases to be written by me?

**Comment by huangqun** [comment]
No, you don't have to write new test cases, but at least your code changes should not break any existing unit tests. That's a strict requirement. I need you to resolve this in the next few hours otherwise we will have to open this for pickup, it's just taking too long.

**Comment by Gauravseta** [comment]
Please dont do this. I will do this.
Can I change .json files of data folder? I know where it is breaking.

**Comment by huangqun** [comment]
What / which .json files?

**Comment by Gauravseta** [comment]
@huangqun request you to review other issues till then, if I fail now I will open it, but pls give me chance, everyone have seen my logic , they will just copy paste mine now with a minor modification in testcases

**Comment by Gauravseta** [comment]
Yesterday, I did fresh checkout of back end repo. Did not add anything of my issue.
Just wrote simple script of alter table add column .
Even that failed.
So test cases will fail every time even if we do minor change as adding column because fields will mismatch

**Comment by Gauravseta** [comment]
Inside data folder there are .json files ,  mentee fields in that .json file of mentees doesn't have my column added so value mismatch occurs.

**Comment by Gauravseta** [comment]
@huangqun m sure u ran alter scripts only after running test data.SQL/ init.SQL ,because if u run alter script before and then test data.SQL or init.SQL it will fail with  that above error

**Comment by Gauravseta** [comment]
I have complete understanding what is going wrong, code is perfect , pls give me sometime I guarantee you it will be completed on time, even issue 315 is ready once this is done , 315 will be done automaticallly I will cover time whatever was wasted here in that 315 issue

**Comment by huangqun** [comment]
No, before merging your fixes all the unit tests pass without any problem. But after your fixes (I have already run the alter scripts) all the unit tests failed. There's definitely something more that needs to be done.

For issue 315 you must also make sure that no unit tests are broken, but apparently I cannot test that one until this is done.

How much more time do you need? I need a timeline, can't keep waiting.

**Comment by Gauravseta** [comment]
Please give me 1 day .please final time. 99% its done. Just a line mismatch is going on .I have realy worked hard for this.please one chance.

**Comment by huangqun** [comment]
That's too much... if we wait that long, we may still not get a valid submission before the challenge ends. I'm sorry but we can't wait that long. We'll open this for pickup and if no one claims it or submits during this time you can continue working on it.

**Comment by huangqun** [comment]
@Gauravseta I have opened it up for pickup for 24h, if no one claims you can claim it again.

I have reminded your here 4 days ago that you need to make sure unit tests all pass, so you really had your chance already: https://github.com/topcoderinc/HPE-LP-MentorMe-Services/pull/79

**Comment by Gauravseta** [comment]
@huangqun ..I found the issue but chq matteo assigned,
Its done I did not find this rule proper.
I was not able to solve ionic deployment issue so I unassigned myself because as u mentioned If one is complete out of line with solution then it is open for pickup.
I agree with that. But here in this issue 99% is done and its made open to pickup now that 1% is done by else and he wins complete amount. Did not find this a good rule by topcoder, the one who did 99% efforts gets wasted and simply termed as fool.
@ward-appirio

**Comment by chq-matteo** [comment]
@Gauravseta if you have fixed the `mvm clean test` issue post the screenshot of the successful tests and screenshots of the language filter feature fixed. In that case I'll unassign myself and let you take the issue. In a couple of hours I'm going to start coding for this, so hurry up.

**Comment by huangqun** [comment]
@Gauravseta I've already explained above, I told you 4 days ago that you should check whether your code changes break the unit tests or not, you had more than enough time to test it and fix any issues if you have tried it earlier. You may not feel good, but that's the rule, you spent so many days on it but didn't even bother to check the unit tests which we explicitly asked you to do, and we can't keep waiting.

Similarly, you have to do the same for your fix for issue 315.

**Comment by Gauravseta** [comment]
@huangqun.. I am in blocked environment now, can't submit PR can't even run anything..I had too less time left hence had to leave from open environment.at that same time it was made open for pickup.
Thats why I told i need sometime not for coding  but to be available..i can submit this at 7 hours from nows.

**Comment by Gauravseta** [comment]
Language filter feature is already fixed since I open the PR first time.
Functionality is completed long back,
Then it was indentation issue of my IDE which took 4 days, finally I have to edit directly in github editor to solve it,
99% done. Now stuck at this basic mvn clean test ,though its done I cannot submit before 8hrs from now.

**Comment by chq-matteo** [comment]
@Gauravseta 8 hours is too much, if my solution ends up using much of your code I'm willing to split the prize with you (a bit)

**Comment by Gauravseta** [comment]
There is nothing left in my code to be added.just test files are to be changed so that mvn clean test runs.
I am not interested in a bit you can keep it with you. No oppose to you ,but admins who just make it open for pickup only because 1% work is left so the actual person doesn't deserve it. Thanks for the offer anyways

**Comment by Gauravseta** [comment]
@huangqun being in extreme need of money, nights are spending in understanding the technology immediately then again more nights are spend in understand the current scenario of code then logic is thought applied tested and then it comes to a good position , and after spending huge time and investing in terms of health too , it is given to some1one else then it hurts, if I would have just taken for a day or two then I wouldn't have minded because I didn't invest much time so it doesn't hurt but after spending 3 nights and it goes away like this. Its painful.

**Comment by huangqun** [comment]
@Gauravseta I don't know how to make it clearer, you have your reasons but I've explained why the decision was made.

You had 4 days just to run a single command (`mvn clean test`) to make sure the changes didn't break the unit tests, if you had spent so much time and effort on this fix why didn't you run the command which would take no more than 5 mins? Let alone I have reminded you to do it.

This is a challenge, if you can't do it in a reasonable timeline then someone else gets the chance to do it, that's how it works. And it will work like this for all future challenges as well, so keep it in mind.

If the fix we accept ends up re-using a lot of the code from your earlier PRs, you'd get a share of the prizes, that's the best we can do.

**Comment by Gauravseta** [comment]
@huangqun  the new PR will be using completely my logic not even a single sentence be changed. Another person can use different variables names,thats it ,optimized flow is what I have done, there are no other way to solve the way I did. Even if a solution exist it will be huge mess unnecessarily.

**Comment by Gauravseta** [comment]
@huangqun I agree I did not run mvn clean test after making changes as it was running OK without modifying column then added my changes tested it and it was working perfect so did not run mvn clean test because till today morning I that mvn clean test runs on main code but 6 hrs back I came to know that there is different code and different validations which shud be passed for successful mvn clean test.
As soon as I understudy what needs to be changed it was open for pickup. And 24hrs I asked not to solve but due to my blocked environment.I can submit new PR 4 hours from now.

**Pull request closed without merging**